### ISSUE TITLE
Represent that two arrays have the same length

### ISSUE THREAD

**Issue by kelloggm** [kelloggm/checker-framework#49]
Consider the following code, which compares two arrays first by length and then lexigraphically:
```
    public int compare(int[] a1, int[] a2) {
      if (a1.length != a2.length) {
        return a1.length - a2.length;
      }
      for (int i = 0; i < a1.length; i++) {
        if (a1[i] != a2[i]) {
          return ((a1[i] > a2[i]) ? 1 : -1);
        }
      }
      return 0;
    }
```

**Comment by kelloggm** [comment]
It would be great if the UBC could figure out that these two arrays have the same length. Should we have a special case for this, though? It shows up in plume-lib several times...

**Comment by kelloggm** [comment]
Possibly the solution to this and several other problems we've run into is to create an analysis that tries to establish which arrays have the same length. Then, anytime we learned something about one, we could learn something about all of them - as we should. Part of MinLen?

**Comment by kelloggm** [comment]
The proposed solution, which I intend to implement, is a new subchecker: the SameLen Checker.

The SameLen checker has three qualifiers: SameLenUnknown, SameLen(arrays), and SameLenBottom. SameLen's argument is a list of array names (i.e. the same as LTLengthOf in the UBC). 

Every array is SameLen to itself. Whenever lengths are checked to be equal, we add the two arrays being checked to each other's SameLen annotations, as well as any other arrays with SameLen to one of the two being checked.

The UBC can then use the SameLen checker as a subchecker. If an array's SameLen type includes one of the arrays listed in a LTLengthOf annotation, then that access is safe.

SameLen suffers from the same soundness problem that the UBC does, and will need to be updated once that problem is solved (i.e. issue 30).

**Comment by smillst** [comment]
Some implementation notes:
>Every array is SameLen to itself.

This should be expressed at `@SameLen("<self>")`.  The Lock Checker has special handling for this expression.  I'll need to generalize that handling and added it to the expression annotation support. (I've already started this as part of changing the Lock Checker to use the new expression annotation support.)

**Comment by kelloggm** [comment]
That's great. This is a fairly high priority issue, though, so I might go ahead and implement it the bad way (i.e. every array keeps track of its own name and viewpoint adaptation just does not work properly) and add an issue to fix it once you've generalized the lock checker's special handling.

**Comment by kelloggm** [comment]
A basic version of the SameLen checker is implemented and operates correctly on this simple test case, which I've checked in as SameLenSimpleTest.java in the upperbound test directory. This is fixed in 3c60995a15fdcfa6738d4feecb1b396e6c785222
